pipelines/_steps/url_preview/upload_image.go

```diff
@@ -21,21 +21,16 @@ func UploadImage(ctx rcontext.RequestContext, image *m.PreviewImage, onHost stri
 	defer image.Data.Close()
 	pr, pw := io.Pipe()
 	tee := io.TeeReader(image.Data, pw)
-	mediaChan := make(chan *database.DbMedia)
-	defer close(mediaChan)
+	mediaChan := make(chan *database.DbMedia, 1)
 	go func() {
 		media, err := pipeline_upload.Execute(ctx, onHost, "", io.NopCloser(tee), image.ContentType, image.Filename, userId, datastores.LocalMediaKind)
 		if err != nil {
 			_ = pw.CloseWithError(err)
 		} else {
 			_ = pw.Close()
 		}
-		go func() {
-			defer func() {
-				recover() // consume write-to-closed-channel errors
-			}()
-			mediaChan <- media
-		}()
+		mediaChan <- media
+		close(mediaChan)
```

Comment:
No, this is a bug. Recovering from a panic does not remedy unclear channel ownership. The go routine for writing is no longer necessary since this is a buffered channel now. And the channel can just be closed afterwards. A simple look at the possible code paths confirms that there is no way the channel could remain unclosed.

Reply:
The panic recovery is indeed a bug, though the `defer close(mediaChan)` needs to be moved up near the channel creation please. The remainder of the change is fine, I think.

Reply:
No. This is a concurrency bug. The channel closing can be deferred in the writing function, but _never_ outside of it. Otherwise, you end with a potential write on a closed channel. And since uploading likely takes a few milliseconds more than thumbnailing (at least on reasonably fast hardware), this will happen more often than not. So the best I could do is:
```go
go func(){
    defer close(mediachan)
    // upload and stuff
```